File: appendixA/src/main/java/org/eternity/variance/Orderer.java
package org.eternity.variance;

public class Orderer {
    private Book book;

    public void order(BookStall bookStall) {
        this.book = bookStall.sell(new IndependentPublisher());
    }

    public static void main(String [] args) {
        new Orderer().order(new BookStall());
        new Orderer().order(new MagazineStore());
    }
}

File: appendixA/src/main/scala/org/eternity/function/variance/Orderer.scala
package org.eternity.function.variance

class Orderer {
  var book: Book = null

  def order(store: IndependentPublisher => Book): Unit = {
    book = store(new IndependentPublisher())
  }
}

object Main extends App {
  new Orderer().order((publisher:IndependentPublisher) => new Book(publisher))
  new Orderer().order((publisher:Publisher) => new Magazine(publisher))
}